fairscale/experimental/nn/mevo.py

```diff
@@ -0,0 +1,483 @@
+    (tokens, d1), (d2, vocabs) = shape
+    assert d1 == d2
+    input = torch.rand(tokens, d1, device=device, dtype=dtype).requires_grad_(True)
+    # Before pytorch 1.9, nn.Linear does not support device and dtype init option. So we use to()
+    # and an if condition.
+    layer = nn.Linear(d2, vocabs, bias=False).to(device)
+    assert dtype in [torch.float16, torch.float32]
+    if dtype == torch.float16:
+        layer = layer.half()
+    weight = layer.weight
+    target = (torch.rand(tokens, device=device) * vocabs).long()
+    return input, weight, target
+
+
+class BaselineSoftmax(nn.Module):
+    """ Baseline softmax that does an output linear projection and a softmax.
+
+        This is intended to be used with an embedding layer with shared weights.
+
+    Args:
+        proj_weight (nn.Parameter):
+            The shared weight.
+        tile_factor (int):
+            Unused. It is here to make kernel init easier with MEVO.
+        log (bool):
+            If True, use log_softmax instead of softmax.
+
+    """
+
+    def __init__(self, proj_weight: nn.Parameter, tile_factor: int = 0, log: bool = True):
+        super().__init__()
+        out_dim, in_dim = proj_weight.shape
+        assert "cuda" in str(proj_weight.device), "weight should be on GPU"
+        self.fc = nn.Linear(in_dim, out_dim, bias=False).to("cuda")
+        assert proj_weight.dtype in [torch.float16, torch.float32]
+        if proj_weight.dtype == torch.float16:
+            self.fc = self.fc.half()
+        self.fc.weight = proj_weight
+        assert self.fc.weight.dtype in [torch.float16, torch.float32], self.fc.weight.dtype
+        self.fp16 = self.fc.weight.dtype == torch.float16
+        self.log = log
+
+    def forward(self, input: torch.Tensor, target: torch.Tensor) -> torch.Tensor:  # type: ignore
+        """ Forward function that computes softmax output with the input and target."""
+        assert isinstance(input, torch.Tensor)
+        assert isinstance(target, torch.Tensor)
+        input, target = _reshape_inputs(input, target)
+        if self.fp16:
+            assert input.dtype == torch.float16
+        x = self.fc(input)
+        # Note that we do softmax in FP32, which is important for numerical stability.
+        if self.log:
+            x = F.log_softmax(x, dim=-1, dtype=torch.float32)
+        else:
+            x = F.softmax(x, dim=-1, dtype=torch.float32)
+        assert x.dtype == torch.float32
+        return x
+
+
+class BaselineSoftmaxNllLoss(BaselineSoftmax):
+    """ Baseline that does an output projection, a softmax & a NLL loss (cross-entropy).
+
+        See BaselineSoftmax above. Constructor is the same. Only difference is in the
+        forward function.
+    """
+
+    def __init__(self, proj_weight: nn.Parameter, tile_factor: int = 0, log: bool = True):
+        super().__init__(proj_weight, tile_factor, log)
+
+    def forward(self, input: torch.Tensor, target: torch.Tensor) -> torch.Tensor:  # type: ignore
```

Comment:
Why do we implement this? is it for testing convenience?

Reply:
correct. This is purely for testing and benchmarking.

Reply:
added a comment.